### PR TITLE
fix issues related to prometheus relabel configs when target allocator is enabled

### PR DESCRIPTION
**Pull request by anunarapureddy**
Related to following issues: #[1623](https://github.com/open-telemetry/opentelemetry-operator/issues/1623), #[1622](https://github.com/open-telemetry/opentelemetry-operator/issues/1622), #[958](https://github.com/open-telemetry/opentelemetry-operator/issues/958)

### Cause:

The issue manifests itself in the following scenarios.

1. First, when the configMap is built by the operator, the unmarshalling logic calls the Prometheus relabel config validation logic ([relabel.go](https://github.com/prometheus/prometheus/blob/3923e83413075a17c5a10739696a879765a05542/model/relabel/relabel.go#L103)), which includes the additional defaults, including the $1 replacement field, appended to the original Prometheus configuration. Upon initialization, the collector pod loads the configMap and interprets $1 as an environment variable, leading to its replacement with an empty string. As a result, the same Prometheus relabel config validation logic fails when the config is unmarshalled in the collector code, as the replacement key is now replaced with an empty string due to env var substitution. 

2. The second scenario occurs when the Prometheus configuration explicitly includes $$ for labelmap, labeldrop, labelkeep or other actions that undergo regex validation in the Prometheus validation code ([relabel.go](https://github.com/prometheus/prometheus/blob/3923e83413075a17c5a10739696a879765a05542/model/relabel/relabel.go#L126)). In this case, when the operator creates the collector configMap, the regex validation fails, causing the configMap to get created with an empty yaml file.

3. In the third scenario, when Prometheus configuration explicitly includes $$ for labelmap, labeldrop, labelkeep or other actions, the TA ConfigMap is created with the $$ signs. Since the TA does not support environment variable substitution, the unmarshalling logic during the ConfigMap mount process in the TA pod will perform regex validation and cause the TA pod creation to fail.

### Resolution:

Below is the guideline from OpenTelemetry documentation:

> Note: Since the collector configuration supports env variable substitution $ characters in your prometheus configuration are interpreted as environment variables. If you want to use $ characters in your prometheus configuration, you must escape them using $$.

Key points considered in order to address the issue:

- To use "$" characters in Prometheus configuration, they must be escaped using "$$".
- The collector configuration supports environment variable substitution, but the TA does not.
- The ConfigMap for the collector pod should have "$$" to prevent env var substitution, unless the user intentionally wants it to happen.
- The TA ConfigMap should have a single "$", as it does not support env var substitution.

The operator was updated to modify the YAML file directly instead of going through the Prometheus unmarshalling logic to update the SD config. The TA ConfigMap creation process was modified to replace "$$" signs with a "$" to address the third scenario. These changes ensure that both the Collector and TA pods are created correctly, even when the Prometheus configuration includes labelmap, labeldrop, labelkeep, or other actions that undergo regex validation.

cc: @alolita @jaronoff97

### CODE REVIEW

**Comment by anunarapureddy** [comment]
@open-telemetry/operator-maintainers As a first-time contributor, I need approval from a maintainer to run workflows. Would someone be able to assist me with obtaining approval?

**Comment by kristinapathak** [comment]
I've looked and thought about this but mostly have follow up questions.

For the issue with including `$1` - my understanding is this is an issue due to the TA being enabled and adding back `$1`'s which are [not supported](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver#getting-started) in the collector configuration:

>Note: Since the collector configuration supports env variable substitution $ characters in your prometheus configuration are interpreted as environment variables. If you want to use $ characters in your prometheus configuration, you must escape them using $$.

While I wish they could be supported, I don't think we should be using the operator to make `$` variables work (which I believe this PR is doing).  My recommendation is instead to modify [`DefaultRelabelConfig`](https://github.com/prometheus/prometheus/blob/3923e83413075a17c5a10739696a879765a05542/model/relabel/relabel.go#L31) to no longer set `Replacement`.  This is not ideal, but because it's a public var, we can set it in the operator code to our desired default config and the prometheus code will use the new value.  Then `$1`'s will no longer be added back into the scrape config when the TA is enabled.

The second issue I see being addressed in this PR is prometheus not validating regex with `$$`.  Is this only with TA enabled or regardless of TA enablement?  And is this due to the code change of `$`->`$$`? I can't find anything in the prometheus code that explains why this wouldn't work.

I looked through the code in the prometheus receiver to find somewhere that `$$` was checked or treated specially but couldn't find anything, so my assumption is that `$$` is considered valid in the scrape config, just doesn't pass the regex checks for some reason.  In other words, the same scrape config with `$$` passed in the operator which causes a problem could be passed in a collector config directly with no problem.  Is that true?  I plan to look into this more but wanted to be sure to provide timely feedback.

My general takeaway is that there are multiple issues here, and I would rather detangle and solve them separately, if possible.  Is there any other issue being addressed here that I've missed?

**Comment by anunarapureddy** [comment]
@open-telemetry/operator-maintainers As a first-time contributor, I need approval from a maintainer to run workflows. Would someone be able to assist me with obtaining approval?

**Comment by anunarapureddy** [comment]
@kristinapathak, I have made changes to the pull request and also updated the description. Can you please review the latest code and provide feedback so that we can iterate on the changes? I was brainstorming with @jaronoff97 and came up with an alternative approach, which is reflected in the updated code.

**Comment by anunarapureddy** [comment]
@kristinapathak, I have made changes to the pull request and also updated the description. Can you please review the latest code and provide feedback so that we can iterate on the changes?

File: pkg/collector/testdata/relabel_config_expected_with_sd_config.yaml
Comment (jaronoff97) [inline pkg/collector/testdata/relabel_config_expected_with_sd_config.yaml:35]: should the `$2` be escaped as `$$2`? This would be what we tell a collector user to do, no?

Comment (jaronoff97) [reply pkg/collector/testdata/relabel_config_expected_with_sd_config.yaml:35]: i.e. does the first `$` escape just `$1` or does it escape the entire `$1_$2`?

File: pkg/collector/reconcile/configmap_test.go
Comment (jaronoff97) [inline pkg/collector/reconcile/configmap_test.go:113]: Why are we removing these lines?

File: pkg/targetallocator/adapters/config_to_prom_config.go
Comment (jaronoff97) [inline pkg/targetallocator/adapters/config_to_prom_config.go:132]: are we unable to just do the replace all on the string configuration we're given? Do we need to only run it on relabel_configs and metric_relabel_configs?

Comment (jaronoff97) [reply pkg/targetallocator/adapters/config_to_prom_config.go:132]: @kristinapathak thinks we should maybe try and just regex this change as well.

File: tests/e2e/targetallocator-features/00-install.yaml
Comment (jaronoff97) [inline tests/e2e/targetallocator-features/00-install.yaml:61]: for completeness, can we also add in a labelkeep action here too?

File: pkg/collector/reconcile/config_replace.go
Comment (jaronoff97) [inline pkg/collector/reconcile/config_replace.go:62]: to confirm my understanding of this – does doing the marshal and then the unmarshal prevent the defaultrelabelconfig from being added and thus removing [this error](https://github.com/prometheus/prometheus/blob/main/model/relabel/relabel.go#LL148C49-L148C49) from being thrown?

Comment (jaronoff97) [reply pkg/collector/reconcile/config_replace.go:62]: or is that going to be in a separate PR?

**Review by jaronoff97 — COMMENTED** [review]
Overall, @kristinapathak and I definitely like this version more than the massaging you had previously. It's easier to understand and the overall design is more extendable. We had a few thoughts/questions to confirm what we had discussed. Thank you again for your contribution 🙇

**Review by jaronoff97 — COMMENTED** [review]
Overall, @kristinapathak and I definitely like this version more than the massaging you had previously. It's easier to understand and the overall design is more extendable. We had a few thoughts/questions to confirm what we had discussed. Thank you again for your contribution 🙇

**Comment by TylerHelmuth** [comment]
@anunarapureddy grab the latest changes from main to deal with the failing autoscale e2e test.

File: pkg/collector/reconcile/config_replace.go
Comment (anunarapureddy) [reply pkg/collector/reconcile/config_replace.go]: @jaronoff97 I had initially planned to handle this separately, but later decided to include it in the same PR to ensure a clear and coherent thought process. I pushed the latest changes to handle this scenario as well.

File: pkg/collector/testdata/relabel_config_expected_with_sd_config.yaml
Comment (anunarapureddy) [reply pkg/collector/testdata/relabel_config_expected_with_sd_config.yaml:35]: @jaronoff97, the code path for constructing the collector configMap does not escape any dollar characters ('$') in the user-specified configuration. Instead, it updates the HTTP SD service configuration by directly modifying the promConfigMap, without the need for unmarshalling and updating the configuration. It is expected that users will properly escape '$' characters in their Prometheus configuration. This approach allows us to support environment variable substitution as well. For instance, in the example above, if a user entered "$$1_$2" in the prometheus configuration, the config map would be created with that value. The "$$1" part will be processed by the regular expression match, while the "$2" part will be replaced by the corresponding environment variable. The fix ensures that there are no problems when the users specify "$$" in the prometheus configuration (as per the OTel guideline).

File: pkg/collector/reconcile/configmap_test.go
Comment (anunarapureddy) [reply pkg/collector/reconcile/configmap_test.go:113]: @jaronoff97 These are additional global defaults that are included in the prometheus configuration when it is unmarshalled. Since we are no longer unmarshalling the configuration, these defaults had to be removed. The [original configuration](https://github.com/open-telemetry/opentelemetry-operator/blob/main/pkg/collector/reconcile/test.yaml) does not contain these defaults.

File: tests/e2e/targetallocator-features/00-install.yaml
Comment (anunarapureddy) [reply tests/e2e/targetallocator-features/00-install.yaml:61]: @jaronoff97 Done!

File: pkg/collector/testdata/relabel_config_expected_with_sd_config.yaml
Comment (jaronoff97) [reply pkg/collector/testdata/relabel_config_expected_with_sd_config.yaml:35]: i see, so in this case `$$1_$2` for a regex match of 'thing' would be `thing_` if $2 is unset? I guess my question is, should we also escape the $2 in our test?

File: pkg/targetallocator/adapters/config_to_prom_config.go
Comment (kristinapathak) [inline pkg/targetallocator/adapters/config_to_prom_config.go:272]: Can you also add an explanation for this in the function comment?

File: pkg/targetallocator/adapters/config_to_prom_config.go
Comment (kristinapathak) [inline pkg/targetallocator/adapters/config_to_prom_config.go:74]: What is the test coverage like on these new functions? It seems like there are a lot of `if` statements present but very few test cases.

**Review by kristinapathak — COMMENTED** [review]
This looks great! Left a few very minor suggestions/questions.

**Comment by jaronoff97** [comment]
@anunarapureddy we just added some more e2e tests to help us catch more of these in the future, and there's already one for labeldropping that you just need to uncomment to enable. ([PR](https://github.com/open-telemetry/opentelemetry-operator/pull/1762)) Could you rebase and uncomment that test when you get a chance as well?

File: pkg/targetallocator/adapters/config_to_prom_config.go
Comment (anunarapureddy) [reply pkg/targetallocator/adapters/config_to_prom_config.go]: Done.

File: pkg/targetallocator/adapters/config_to_prom_config.go
Comment (anunarapureddy) [reply pkg/targetallocator/adapters/config_to_prom_config.go:281]: Done.

File: pkg/targetallocator/adapters/config_to_prom_config.go
Comment (anunarapureddy) [reply pkg/targetallocator/adapters/config_to_prom_config.go:83]: Adding test cases for missing and invalid configurations as well.

File: pkg/targetallocator/adapters/config_to_prom_config.go
Comment (anunarapureddy) [reply pkg/targetallocator/adapters/config_to_prom_config.go:222]: Done.

**Comment by anunarapureddy** [comment]
@jaronoff97 Uncommented the "labeldrop" e2e test case and added additional unit test cases to cover missing or invalid configurations.

**Comment by jaronoff97** [comment]
Awesome, thank you! running those now and will give another review today as well once they pass 🤞

**Comment by jaronoff97** [comment]
looks like a single linter bug @anunarapureddy

**Comment by anunarapureddy** [comment]
@jaronoff97 Seems like the linter gave up because it took too long. I checked the code on my own machine, and I didn't spot any lint failures. Everything looks good from my end!

File: pkg/collector/reconcile/config_replace.go
Comment (jaronoff97) [inline pkg/collector/reconcile/config_replace.go:41]: i think we still need this? otherwise we'll run in to the unmarshal error about an unknown field.

**Review by jaronoff97 — APPROVED** [review]
just one quick question, otherwise this is looking good to me!

File: pkg/collector/reconcile/config_replace.go
Comment (anunarapureddy) [reply pkg/collector/reconcile/config_replace.go]: @jaronoff97 Apologies! It seems that this happened during the rebase and conflict resolution process. I have thoroughly reviewed everything now, and it appears to be in good shape.

File: pkg/targetallocator/adapters/config_to_prom_config.go
Comment (swiatekm-sumo) [inline pkg/targetallocator/adapters/config_to_prom_config.go:107]: I don't think we should error if these don't exist, but rather just exit. Target allocator doesn't require Prometheus configuration.

Maybe this should be done in a separate PR though, as I believe this mirrors the current behaviour. WDYT @jaronoff97 ?

File: pkg/collector/reconcile/configmap.go
Comment (swiatekm-sumo) [inline pkg/collector/reconcile/configmap.go:106]: I think it's worth adding a comment here as well. Without context it's not clear at all why we'd need to unescape these here.

File: pkg/targetallocator/adapters/config_to_prom_config.go
Comment (swiatekm-sumo) [inline pkg/targetallocator/adapters/config_to_prom_config.go:200]: Should we error if the Prometheus config doesn't exist, or just return quietly? See my comment for the TA config unescaping as well.

**Review by swiatekm-sumo — COMMENTED** [review]
Looks good to me overall. I think we're being too strict about requiring the Prometheus config to exist, but that can be relaxed in a future PR if need be. Overall, despite the verbosity (should we consider parsing permissively into structs instead?), I think this will greatly improve the reliability of the config transformations.

File: pkg/targetallocator/adapters/config_to_prom_config.go
Comment (anunarapureddy) [reply pkg/targetallocator/adapters/config_to_prom_config.go:107]: @jaronoff97 @swiatekm-sumo I can create a follow up PR to relax this restriction.

File: pkg/collector/reconcile/configmap.go
Comment (anunarapureddy) [reply pkg/collector/reconcile/configmap.go]: Done. Added a comment with the reason to unescape.

**Comment by anunarapureddy** [comment]
@jaronoff97 @swiatekm-sumo Addressed the review comments and I can work on creating a follow up PR to relax the restrictions on prometheus configuration.

**Review by swiatekm-sumo — APPROVED** [review]
LGTM

**Comment by anunarapureddy** [comment]
@bryan-aguilar That's a good suggestion. I can create a follow up PR to update the TA documentation.

**Comment by bryan-aguilar** [comment]
> @bryan-aguilar That's a good suggestion. I can create a follow up PR to update the TA documentation.

Do we predict further changes to this functionality after this PR? I worry about documentation drift and a possible loss of context if we don't lump documentation and functional changes in the same PR.

**Comment by anunarapureddy** [comment]
> > @bryan-aguilar That's a good suggestion. I can create a follow up PR to update the TA documentation.
> 
> Do we predict further changes to this functionality after this PR? I worry about documentation drift and a possible loss of context if we don't lump documentation and functional changes in the same PR.

Could you please review the Updated the README documentation which reflects the translations being done?

File: README.md
Comment (Aneurysm9) [inline README.md:450]: Shouldn't these values be preserved in the configuration given to the collector?  These are scrape-time options that are needed by the receiver.

File: README.md
Comment (Aneurysm9) [inline README.md:472]:
```suggestion
Note how the Operator removes any existing service discovery configurations (e.g., `static_configs`, `file_sd_configs`, etc.) from the `scrape_configs` section and adds an `http_sd_configs` configuration pointing to a Target Allocator instance it provisioned.
```

More than one `http_sd_configs` entry will be present if there is more than one scrape job in the incoming config.  There should be one per job.  The example only had one job, but a real config is likely to have more.

File: pkg/collector/reconcile/config_replace.go
Comment (Aneurysm9) [inline pkg/collector/reconcile/config_replace.go:75]: I don't believe it is safe to assume that the receiver ID is always `prometheus`.  There may be multiple prometheus receiver configurations and any or all of them may have a name in addition to the type in their identifier.  This needs to iterate over all entries in the `receivers` map and process any that have an identifier matching `/^prometheus(\/.+)?$/`.

File: README.md
Comment (anunarapureddy) [reply README.md:472]: @Aneurysm9 The replacement process occurs for each job in the incoming configuration. The code iterates through all the jobs and updates the configuration accordingly. If it would be helpful for clarity, I can include an additional job in the example.

File: pkg/collector/reconcile/config_replace.go
Comment (anunarapureddy) [reply pkg/collector/reconcile/config_replace.go:75]: @Aneurysm9 The assignment of this logic remains unchanged as it is an existing implementation (https://github.com/open-telemetry/opentelemetry-operator/blob/d6d9f2f628a80858460cddbbee85b4ca9f13fecf/pkg/collector/reconcile/config_replace.go#L99). However, I can certainly address it as an improvement in a follow-up PR.

**Review by kristinapathak — APPROVED** [review]
This is awesome! Thanks for putting in so much time and effort into getting `$`s working 🙂

**Review by bryan-aguilar — APPROVED** [review]
Thank you for adding the documentation!

**Review by jaronoff97 — APPROVED** [review]

**Comment by jaronoff97** [comment]
once tests pass, will merge this in :)

**Pull request merged**
